privacyidea_radius.pm

```diff
@@ -414,6 +418,15 @@ sub authenticate {
     }
     if ( exists( $RAD_REQUEST{'User-Password'} ) ) {
         my $password = $RAD_REQUEST{'User-Password'};
+        if(my $encodingname = Encode::Detect::Detector::detect($password )){
+          &radiusd::radlog( Debug, "Password Encoding: $encodingname" );
+          my $decoded = decode $encodingname, $password ;
+          &radiusd::radlog( Debug, "Password Decoded: $decoded" );
+          $RAD_REQUEST{'User-Password'} = $decoded;
```

Comment:
Since the NPS sits befpre the FreeRADIUS I do not see the necissity to write the new decoded passwort into the RAD_REQUEST. Why would it?

If there is no reason, I would remove this line. Thx.

Reply:
We use PrivacyIDEA with a push token for VPN Connections.
So the User is using his normal AD Credentials.
The NPS Server checks the password first. After that it is relayed to PrivacyIDEA.
It does check the password too.
When PrivacyIDEA uses the wrong encoded password the authentification fails.

Reply:
Yes, but I think there is no need to write the modified password back to the orinigal RADIUS request.
So the line 425 can be omitted.

Reply:
Yes you are right. I tested it again and it works if I just overwrite the password.

I tested my code again found that it fails with a few special Passwords.
I used the Password N0cßÜÖÄü2019 for testing.

I used this code to test  https://gist.github.com/marcohald/21dbdb5f1ba3dbf28f50964a76758142
with this output:

```
rlm_perl: Password Encoding: KOI8-R
rlm_perl: fromto KOI8-R UTF-8: N0c?????2019
rlm_perl: fromto iso-8859-1 UTF-8: N0cßÜÖÄü2019
```

I don't know why it is recognized as "KOI8-R" other Passwords are recognized as "windows-1255" but the decoding from it looks wrong and don't work.

```
rlm_perl: Password Encoding: windows-1255
rlm_perl: fromto windows-1255 UTF-8: Test12345678???
rlm_perl: fromto iso-8859-1 UTF-8: Test12345678öäü

```
I used the "iso-8859-1" encoding as static string and this seems to be working.
But i don't know what is the right way to check if it is encoded.

Reply:
Maybe you simply need to know, that the password is encoded. After all I think it is just a bytestring that arrives at freeradius and it is a matter of the interpretation (encoding) what this bytestring is supposed to mean.

The ``User-Password`` is actually encrypted with the shared secret. So if you choose the wrong shared secret, you will end up with gibberish. ...or is it a really strange encoding?

So one way could be to actually configure the encoding in ``rlm_perl.ini``. I.e. if you know that your passwords are *windows-1255* encoding, we could add a config like

~~~~
[Default]
ENCODING = windows-1255
~~~~

Reply:
I think that this is a good way to configure it in the rlm_perl.ini 
Because other Services may use other encodings.
It seems that the encoding for Windows NPS is iso-8859-1
Is it necessary to check if the string needs decoding or can we simply call it every time when the Encoding is set in the ini file?
Could you do the implementation?